File: kotlin/src/main/kotlin/dev/wogan/advent/kotlin/year2024/Day01.kt
package dev.wogan.advent.kotlin.year2024

import dev.wogan.advent.kotlin.partitionOn
import dev.wogan.advent.kotlin.readInput
import dev.wogan.advent.kotlin.separate
import kotlin.math.abs

val whitespace = "\\s+".toRegex()

fun wellFormed(input: List<String>): Pair<List<Int>, List<Int>> =
    input.map {
        val (a, b) = it.partitionOn(whitespace)
        a.toInt() to b.toInt()
    }.separate()

fun part1(input: List<String>): Int {
    val (a, b) = wellFormed(input)
    return a.sorted().zip(b.sorted()).sumOf { (a, b) -> abs(a - b) }
}

fun part2(input: List<String>): Int {
    val (a, b) = wellFormed(input)
    val map = b.groupingBy { it }.eachCount()
    return a.sumOf { it * (map[it] ?: 0) }
}

fun main() {
    val input = readInput(1, 2024)
    println("Part 1: ${part1(input)}")
    println("Part 2: ${part2(input)}")
}

File: scala/src/main/scala/year2024/Day01.scala
package dev.wogan.advent.scala
package year2024

import cats.effect.IO
import cats.syntax.all.*
import fs2.Stream

object Day01 extends Day(1) {

  val parser = Parsers.int ~ (Parsers.whitespace *> Parsers.int)

  def separate(input: Input): (Stream[IO, Int], Stream[IO, Int]) =
    input.flatMap(parser.parseAllStream).unzip

  override def part1(input: Input): Output =
    val (a, b) = separate(input)
    a.sorted.zip(b.sorted).map((a, b) => Math.abs(a - b)).sum

  override def part2(input: Input): Output =
    val (a, b) = separate(input)
    val scores = for
      map <- Stream.eval(b.compile.toList.map(_.groupMapReduce(identity)(_ => 1)(_ + _)))
      i <- a
    yield i * map.getOrElse(i, 0)
    scores.sum

}
